lib/src/main/java/oiccli/Service.java

```diff
@@ -0,0 +1,443 @@
+
+        if (requestArgs == null) {
+            requestArgs = new HashMap<>();
+        }
+
+        List<Map<String, Object>> returnedArgs = this.doPreConstruct(clientInfo, requestArgs, args);
+
+        if (!this.msgType.c_param.containsKey("state")) {
+            args.remove("state");
+        }
+
+        Map<String, Object> argsParam = this.gatherRequestArgs(clientInfo, requestArgs);
+        this.msgType(argsParam);
+
+        return this.doPostConstruct(clientInfo, requestArgs, returnedArgs.get(1));
+    }
+
+    public String getEndpoint(Map<String, String> args) throws MissingEndpoint {
+        String uri = args.get("endpoint");
+
+        if (uri != null) {
+            args.remove("endpoint");
+        } else {
+            if (StringUtil.isNotNullAndNotEmpty(this.endpoint)) {
+                uri = this.endpoint;
+            } else {
+                throw new MissingEndpoint("No endpoint specified");
+            }
+        }
+
+        return uri;
+    }
+
+    public Map<String, Object> uriAndBody(Message cis, String method, Map<String, String> args) throws UnsupportedEncodingException, UnsupportedType, MissingEndpoint {
+        String uri = this.getEndpoint(args);
+        Map<String, Object> response = Util.getOrPost(uri, method, cis, args);
+        response.put("cis", cis);
+        Map<String, Object> hMap = new HashMap<>();
+        hMap.put("headers", response.get("headers"));
+        response.put("hArgs", hMap);
+
+        return response;
+    }
+
+    public Map<String, String> initAuthenticationMethod(Message cis, ClientInfo clientInfo, String authenticationMethod,
+                                                        Map<String, Object> httpArgs, Map<String, String> args) {
+        return initAuthenticationMethod(cis, clientInfo, authenticationMethod, httpArgs, null, args);
+    }
+
+    public Map<String, String> initAuthenticationMethod(Message cis, ClientInfo clientInfo, String authenticationMethod,
+                                                        Map<String, Object> requestArgs, Map<String, String> httpArgs, Map<String, String> args) {
+        if (httpArgs == null) {
+            httpArgs = new HashMap<>();
+        }
+
+        if (StringUtil.isNotNullAndNotEmpty(authenticationMethod)) {
+            //return this.client_authn_method[authn_method]().construct(
+            //      cis, cli_info, request_args, http_args, **kwargs);
+        } else {
+            return httpArgs;
+        }
+    }
+
+    public Map<String, Object> requestInfo(ClientInfo clientInfo, String method, Map<String, Object> requestArgs,
+                                           String bodyType, String authenticationMethod, boolean lax, Map<String, String> args) throws NoSuchFieldException, IllegalAccessException, MissingEndpoint, UnsupportedEncodingException, UnsupportedType {
+        if (method == null) {
```

Comment:
I don't like all these null checks. I think it's a bad practice to have "optional parameters" in the signature. (i.e. calling a method like `method(123, true, null, null, null, "something", null);`). We should at least know what values a method can take.

Reply:
In oicmsg, those are the default values, so I decided to overload those methods.